lib/mobility-core/src/Kernel/Storage/Esqueleto/Transactionable.hs

```diff
@@ -55,48 +68,99 @@ instance {-# OVERLAPPING #-} Transactionable' SqlDB m => Transactionable' SqlDB
 -- But with INCOHERENT it will always use Transactionable' m instance.
 -- It's fine since Transactionable' SqlDB instance should be used only in find... functions,
 -- which have proper Transactionable' instance.
-instance {-# INCOHERENT #-} (HasEsqEnv m r, MonadThrow m, Log m) => Transactionable' SqlDB m where
+instance {-# INCOHERENT #-} (HasEsqEnv m r, MonadThrow m, Log m, Finalize m) => Transactionable' SqlDB m where
   runTransaction m = do
     dbEnv <- asks (.esqDBEnv)
-    runTransactionImpl dbEnv m
+    (a, _env) <- runTransactionImpl dbEnv m
+    pure a
+
+-- | Type safe plug for ignoring finalize function
+ignoreFinalize :: forall m. Finalize m => (m () -> SqlDB ()) -> (m () -> SqlDB ())
+ignoreFinalize _ _ = pure ()
+
+-- This function created for type safety. Finalize function should be applied to the same monad within transaction
+runTransactionF :: forall m r a. (HasEsqEnv m r, MonadThrow m, Log m, Finalize m) => ((m () -> SqlDB ()) -> SqlDB a) -> m a
+runTransactionF mkAction = do
+  dbEnv <- asks (.esqDBEnv)
+  let m = mkAction (finalizeImpl @m)
+  (result, env) <- runTransactionImpl dbEnv m
+  applyFinalizeImpl @m env
+  pure result
+
+-- | This function should not be used outside of Transactionable module!
+finalizeImpl :: forall m. Finalize m => m () -> SqlDB ()
+finalizeImpl someAction = do
+  SqlDBEnv {..} <- get
+  let mbPrevActions = cast @_ @(m ()) actions
```

Comment:
Is it possible to remove cast & Typeable instance from the whole thing? We won't use it without runTransactionF and monad  will be decided with first finalize call anyway

Reply:
> Is it possible to remove cast & Typeable instance from the whole thing? We won't use it without runTransactionF and monad will be decided from the start

I don't know how its possible, If you know please add your variant here. In previous solution by @pavelzarubin he didn';t use cast & Typeable instance. 
https://github.com/nammayatri/nammayatri/pull/67
But therefore he needed to make `SqlDB` type polimorphic (`SqlDB m a`). It required a lot of changes all over the code, so my solution I like better.

Reply:
Another my solution was to create some local type synonyms in each app, like:

`type SqlDB a = Esq.SqlDB Flow a`

but its not working with polymorphic code, when we don't know exact type of `m`